src/catalog/src/remote/manager.rs

```diff
@@ -254,15 +256,44 @@ impl RemoteCatalogManager {
     ) -> Result<()> {
         info!("initializing tables in {}.{}", catalog_name, schema_name);
         let mut table_num = 0;
-        let mut tables = self.iter_remote_tables(catalog_name, schema_name).await;
-        while let Some(r) = tables.next().await {
-            let (table_key, table_value) = r?;
-            let table_ref = self.open_or_create_table(&table_key, &table_value).await?;
-            schema.register_table(table_key.table_name.to_string(), table_ref)?;
-            info!("Registered table {}", &table_key.table_name);
-            max_table_id = max_table_id.max(table_value.table_id());
-            table_num += 1;
+        let tables = self.iter_remote_tables(catalog_name, schema_name).await;
+        let kvs = tables.collect::<Vec<_>>().await;
+        let node_id = self.node_id;
+
+        let mut joins = kvs
+            .into_iter()
+            .map(|kv| {
+                let (table_key, table_value) = kv?;
+                let engine_manager = self.engine_manager.clone();
+                Ok(tokio::spawn(async move {
+                    let table_ref =
+                        open_or_create_table(node_id, engine_manager, &table_key, &table_value)
+                            .await?;
+                    Ok(table_ref)
+                }))
+            })
+            .collect::<Result<Vec<_>>>()?;
+
+        while !joins.is_empty() {
+            match futures::future::select_all(joins).await {
+                (Ok(join_re), _, remaining) => {
+                    joins = remaining;
+                    let table_ref = join_re?;
+                    let table_info = table_ref.table_info();
+
+                    let table_name = &table_info.name;
+                    let table_id = table_info.ident.table_id;
+                    schema.register_table(table_name.clone(), table_ref)?;
+                    info!("Registered table {}", table_name);
+                    max_table_id = max_table_id.max(table_id);
+                    table_num += 1;
+                }
+                (Err(source), _, _) => {
+                    return Err(ParallelOpenTable { source });
+                }
+            }
         }
```

Comment:
This can be simplified like:
```suggestion
        let kvs = tables.try_collect::<Vec<_>>().await?;

        let node_id = self.node_id;

        let mut joins = kvs
            .into_iter()
            .map(|(table_key, table_value)| {
                let engine_manager = self.engine_manager.clone();
                tokio::spawn(async move {
                    open_or_create_table(node_id, engine_manager, &table_key, &table_value).await
                })
            })
            .collect::<Vec<_>>();

        let vec = futures::future::join_all(joins).await;

        for res in vec {
            let table_ref = res.context(ParallelOpenTableSnafu)??;
            let table_info = table_ref.table_info();
            let table_name = &table_info.name;
            let table_id = table_info.ident.table_id;
            schema.register_table(table_name.clone(), table_ref)?;
            info!("Registered table {}", table_name);
            max_table_id = max_table_id.max(table_id);
            table_num += 1;
        }
```